crates/ruma-appservice-api/src/event/push_events.rs

```diff
@@ -2,11 +2,213 @@
+use std::collections::BTreeMap;
+
+/// Type for passing ephemeral data to homeservers.
+#[derive(Clone, Debug, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+#[serde(tag = "type", content = "content")]
+pub enum Edu {
+    /// An EDU representing presence updates for users of the sending homeserver.
+    #[serde(rename = "m.presence")]
+    Presence(PresenceContent),
+
+    /// An EDU representing receipt updates for users of the sending homeserver.
+    #[serde(rename = "m.receipt")]
+    Receipt(ReceiptContent),
+
+    /// A typing notification EDU for a user in a room.
+    #[serde(rename = "m.typing")]
+    Typing(TypingContent),
+
+    #[doc(hidden)]
+    _Custom(JsonValue),
+}
+
+#[derive(Debug, Deserialize)]
+#[cfg(feature = "unstable-msc2409")]
+struct EduDeHelper {
+    /// The message type field
+    r#type: String,
+    content: Box<RawJsonValue>,
+}
+
+#[cfg(feature = "unstable-msc2409")]
+impl<'de> Deserialize<'de> for Edu {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let EduDeHelper { r#type, content } = from_raw_json_value(&json)?;
+
+        Ok(match r#type.as_ref() {
+            "m.presence" => Self::Presence(from_raw_json_value(&content)?),
+            "m.receipt" => Self::Receipt(from_raw_json_value(&content)?),
+            "m.typing" => Self::Typing(from_raw_json_value(&content)?),
+            _ => Self::_Custom(from_raw_json_value(&content)?),
+        })
+    }
+}
+
+/// The content for "m.presence" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+pub struct PresenceContent {
+    /// A list of presence updates that the receiving server is likely to be interested in.
+    pub push: Vec<PresenceUpdate>,
+}
+
+#[cfg(feature = "unstable-msc2409")]
+impl PresenceContent {
+    /// Creates a new `PresenceContent`.
+    pub fn new(push: Vec<PresenceUpdate>) -> Self {
+        Self { push }
+    }
+}
+
+/// An update to the presence of a user.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[cfg(feature = "unstable-msc2409")]
+pub struct PresenceUpdate {
+    /// The user ID this presence EDU is for.
+    pub user_id: OwnedUserId,
+
+    /// The presence of the user.
+    pub presence: PresenceState,
+
+    /// An optional description to accompany the presence.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub status_msg: Option<String>,
+
+    /// The number of milliseconds that have elapsed since the user last did something.
+    pub last_active_ago: UInt,
+
+    /// Whether or not the user is currently active.
+    ///
+    /// Defaults to false.
+    #[serde(default)]
```

Comment:
`serde(default)` should always be accompanied with `serde(skip_serializing_if)`. You can use `ruma_common::serde::is_default` here.